proposals/0257-string-uninitialized-initializer.md

```diff
@@ -0,0 +1,141 @@
+  ///     // Prints "Optional(Caf�)"
+  ///
+  ///     let s = String(unsafeUninitializedCapacity: invalidUTF8.count,
+  ///                    initializingUTF8With: { (ptr, count) in
+  ///         ptr.initializeFrom(invalidUTF8)
+  ///         count = 0
+  ///     })
+  ///     // Prints "Optional("")"
+  ///
+  /// - Parameters:
+  ///   - capacity: The number of UTF-8 code units worth of memory to allocate
+  ///       for the String.
+  ///   - initializer: A closure that initializes elements and sets the count of
+  ///       the new String
+  ///     - Parameters:
+  ///       - buffer: A buffer covering uninitialized memory with room for the
+  ///           specified number of UTF-8 code units.
+  ///       - initializedCount: Set this to the number of elements in `buffer`
+  ///           that were actually initialized by the `initializer`
+  @inlinable @inline(__always)
+  public init(
+    unsafeUninitializedCapacity capacity: Int,
+    initializingUTF8With initializer: (
+    _ buffer: UnsafeMutableBufferPointer<UInt8>,
+    _ initializedCount: inout Int
+    ) throws -> Void
+  ) rethrows
+```
+
+### Specifying a capacity
+
+The initializer takes the specific capacity that a user wants to work with as a
+parameter. The buffer passed to the closure has a count that is exactly the
+same as the specified capacity, even if the ultimate size of the new `String` is larger.
+
+### Guarantees after throwing
+
+Because `UTF8.CodeUnit` is a trivial type, there are no special considerations about the state of the buffer when an error is thrown, unlike `Array`.
+
+## Source compatibility
+
+This is an additive change to the standard library,
+so there is no effect on source compatibility.
+
+## Effect on ABI stability
+
+The new initializer will be part of the ABI, and will result in calls to a new @usableFromInline symbol being inlined into client code. Use of the new initializer is gated by @availability though, so there's no back-deployment concern.
+
+## Effect on API resilience
+
+The additional APIs will be a permanent part of the standard library,
+and will need to remain public API. 
+
+## Alternatives considered
+
+### Returning the new count from the initializer closure
+
+This is more plausible for `String` than it was for `Array`, since there's no need to deal with deinitialization of elements in partially initialized buffers (`UInt8`s are trivial). However, it was considered more valuable to match `Array`'s behavior here.
+
+### Returning a `Bool` to indicate success from the closure
+
+Requiring people to either `throw` or check in the caller for an empty `String` return if the initializing closure fails is slightly awkward, but making the initializer failable is at least as awkward, and would be inconsistent with `Array`.
+
+### Validating UTF-8 instead of repairing invalid UTF-8
+
+Matching the behavior of most other `String` initializers here also makes it more ergonomic to use, since it can be non-failable this way.
```

Comment:
You can also say that the default behavior for String initializers is to do repairing, delivering the benefit of non-failable initializers and performance advantages of ahead-of-time validation. Should String flesh out the validation story with non-repairing failable alternatives (or even non-failable unsafe unchecked alternatives), uninitialized access APIs can also be added.

Reply:
Makes sense

Reply:
only to understand:
why cannot be failable in symmetry  with (for example..)

init?(utf8String: UnsafePointer<CChar>)